machine/xen/v1alpha1.go

```diff
@@ -0,0 +1,536 @@
+
+	machinev1alpha1 "kraftkit.sh/api/machine/v1alpha1"
+	"kraftkit.sh/config"
+	"kraftkit.sh/internal/logtail"
+	"kraftkit.sh/log"
+	"kraftkit.sh/machine/network/macaddr"
+	"kraftkit.sh/unikraft/export/v0/ukargparse"
+	"kraftkit.sh/unikraft/export/v0/uknetdev"
+	"kraftkit.sh/unikraft/export/v0/vfscore"
+	"xenbits.xenproject.org/git-http/xen.git/tools/golang/xenlight"
+)
+
+const (
+	XenMemoryScale   = 1024
+	XenMemoryDefault = 64
+	XenCPUsDefault   = 1
+)
+
+type machineV1alpha1Service struct{}
+
+func NewMachineV1alpha1Service(ctx context.Context) (machinev1alpha1.MachineService, error) {
+	return &machineV1alpha1Service{}, nil
+}
+
+func (service *machineV1alpha1Service) Create(ctx context.Context, machine *machinev1alpha1.Machine) (*machinev1alpha1.Machine, error) {
+	cfg, err := NewXenConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	if machine.Status.KernelPath == "" {
+		return machine, fmt.Errorf("cannot create xen instace without a kernel")
+	}
+
+	if !slices.Contains([]string{"arm64", "x86_64", "amd64", "arm"}, machine.Spec.Architecture) {
+		return machine, fmt.Errorf("unsupported architecture %s", machine.Spec.Architecture)
+	}
+
+	if _, err := os.Stat(machine.Status.KernelPath); err != nil && os.IsNotExist(err) {
+		return machine, fmt.Errorf("supplied kernel path does not exist: %s", machine.Status.KernelPath)
+	}
+
+	if machine.ObjectMeta.UID == "" {
+		machine.ObjectMeta.UID = uuid.NewUUID()
+	}
+
+	machine.Status.State = machinev1alpha1.MachineStateUnknown
+
+	if len(machine.Status.StateDir) == 0 {
+		machine.Status.StateDir = filepath.Join(config.G[config.KraftKit](ctx).RuntimeDir, string(machine.ObjectMeta.UID))
+	}
+
+	if err := os.MkdirAll(machine.Status.StateDir, fs.ModeSetgid|0o775); err != nil {
+		return machine, err
+	}
+
+	if len(machine.Status.LogFile) == 0 {
+		machine.Status.LogFile = filepath.Join(machine.Status.StateDir, "machine.log")
+	}
+
+	fd, err := os.Create(machine.Status.LogFile)
+	if err != nil {
+		return machine, err
+	}
+	defer fd.Close()
+
+	if machine.Spec.Resources.Requests == nil {
+		machine.Spec.Resources.Requests = make(corev1.ResourceList, 2)
+	}
+
+	if machine.Spec.Resources.Requests.Memory().Value() == 0 {
+		quantity, err := resource.ParseQuantity(fmt.Sprintf("%d", XenMemoryDefault) + "Mi")
+		if err != nil {
+			machine.Status.State = machinev1alpha1.MachineStateFailed
+			return machine, err
+		}
+
+		machine.Spec.Resources.Requests[corev1.ResourceMemory] = quantity
+	}
+
+	if machine.Spec.Resources.Requests.Cpu().Value() == 0 {
+		quantity, err := resource.ParseQuantity(fmt.Sprintf("%d", XenCPUsDefault))
+		if err != nil {
+			machine.Status.State = machinev1alpha1.MachineStateFailed
+			return machine, err
+		}
+
+		machine.Spec.Resources.Requests[corev1.ResourceCPU] = quantity
+	}
+
+	cfg.BInfo.MaxVcpus = int(machine.Spec.Resources.Requests.Cpu().Value())
+	cfg.BInfo.MaxMemkb = uint64(machine.Spec.Resources.Requests.Memory().Value() / XenMemoryScale)
```

Comment:
A corner case, but, if you want to start a machine with `<1 Ki` this division will make it have 0 

Can you do a check above to see, if memory was specified, that it's at least `1KiB` :sweat_smile: